Add support for buildpack test command

dzaporozhets opened the pull request:
Attempt to add support for `buildpack test` command. 

For https://github.com/gliderlabs/herokuish/issues/257

dzaporozhets commented:
@michaelshobbs I am quite new to herokuish and rarely do any bash coding. I am wondering if someone from maintainers can help me turn this into mergeable contribution.

dzaporozhets commented:
`buildpack-test` is mostly copied code from `buildpack-execute` with small modification.  So far `test-compile` works well but `test` does fail:

```
→ make build-in-docker
→ docker run --rm -v /Users/dzaporozhets/Projects/ruby-sinatra:/tmp/app herokuish:dev /bin/herokuish buildpack test
::: BUILDING APP :::
-----> Ruby app detected
::: PREPARE FOR RUNNING TESTS :::
-----> Setting up Test for Ruby/Rack
-----> Using Ruby version: ruby-2.3.3
-----> Installing dependencies using bundler 1.13.7
       Running: bundle install --without development --path vendor/bundle --binstubs vendor/bundle/bin -j4 --deployment
       Warning: the running version of Bundler (1.13.7) is older than the version that created the lockfile (1.15.3). We suggest you upgrade to the latest version of Bundler by running `gem install bundler`.
       Fetching gem metadata from http://rubygems.org/........
       Fetching version metadata from http://rubygems.org/.
       Installing diff-lcs 1.3
       Installing rake 12.0.0
       Installing rack 1.5.2
       Using bundler 1.13.7
       Installing rspec-support 3.6.0
       Installing rack-protection 1.5.0
       Installing rspec-core 3.6.0
       Installing tilt 1.3.6
       Installing rspec-expectations 3.6.0
       Installing rspec-mocks 3.6.0
       Installing rspec 3.6.0
       Installing sinatra 1.4.2
       Bundle complete! 4 Gemfile dependencies, 12 gems now installed.
       Gems in the group development were not installed.
       Bundled gems are installed into ./vendor/bundle.
       Bundle completed (9.79s)
       Cleaning up the bundler cache.
       Warning: the running version of Bundler (1.13.7) is older than the version that created the lockfile (1.15.3). We suggest you upgrade to the latest version of Bundler by running `gem install bundler`.
-----> Detecting rake tasks

::: RUNNING TESTS :::
-----> Running test: bundle exec rspec
       bash: bundle: command not found
```

I used https://github.com/dzaporozhets/ruby-sinatra as sample ruby app to test functionality

michaelshobbs commented:
@dzaporozhets thanks for the contribution. we will certainly need some test coverage and documentation updates in order to get this PR merged.

dzaporozhets commented:
@michaelshobbs sure. At this point, I would like just to make it work. I would appreciate if you could take a look at the code and give me some hints on why it fails on https://github.com/gliderlabs/herokuish/pull/271/files#diff-57c992bbac59903eb3fa272073dc4c8eR222

michaelshobbs commented:
DISCLAIMER: I am by no means a ruby expert nor do I deploy/run/manage ruby apps.

Based on the output and at the risk of sounding obvious, it's failing due to `bundle: command not found`. If I were you, I'd start by sorting out where (if anywhere) `bundle` exists in the container.

michaelshobbs commented on include/buildpack.bash:
```diff
@@ -176,3 +176,54 @@ buildpack-execute() {
+	[[ "$USER" ]] || randomize-unprivileged
+	buildpack-setup > /dev/null
+
+	if [[ -n "$BUILDPACK_URL" ]]; then
```
There is some code duplication here. I'd rather see some of this broken out into a function and get re-used by this and anywhere else we handle buildpack detection.

dzaporozhets replied:
@michaelshobbs yep I will extract this into `select-buildpack` function

dzaporozhets commented on include/buildpack.bash:
```diff
@@ -176,3 +176,54 @@ buildpack-execute() {
+	shopt -s dotglob nullglob
+	rm -rf $app_path/*
+	mv $build_path/* $app_path
+	shopt -u dotglob nullglob
```
@michaelshobbs I copied 5 lines above from `buildpack-execute`.  Any hints what it used for?

michaelshobbs replied:
This moves the compiled app into the `app_path`. I don't think it's necessary for the testing use case

dzaporozhets replied:
cool, thanks

dzaporozhets commented:
> based on the output and at the risk of sounding obvious, it's failing due to bundle: command not found. If I were you, I'd start by sorting out where (if anywhere) bundle exists in the container.

@michaelshobbs yeah, sorry my communication was unclear :) Basically my question is next:  Do
I need extra code to execute buildpacks commands one by one or just running `$selected_path/bin/CMD` should be enough.

michaelshobbs commented:
AFAIK, `$selected_path/bin/CMD` is sufficient. However, since we allow custom buildpacks, we should probably wrap these in a check if the command exists and is executable. If not, pop an error and bail.

dzaporozhets commented:
I refactored it a bit. Tested on Ruby and Clojure projects. 

So far: 
* `test-compile` works
* `test` does not work

dzaporozhets commented:
Ok seems `test` does not work because you need to source files from `.profile.d` directory created by `test-compile`.

dzaporozhets commented:
@michaelshobbs I made it work and did some refactoring to avoid code duplication. Can you please take a look at the code base while I try to figure out how to write tests for it?

dzaporozhets commented:
Also, one question to decide on: are we ok having `buildpack test` command that runs both `test-compile` and `test` inside? Or we can split those into 3 commands: 

1. `buildpack test-compile`
2. `buildpack test`
3. `buildpack test-with-compile`

For personal use, I don't think separation makes sense but there is a potential for caching between compile and test if you build some heavy used CI based on herokuish.   

@michaelshobbs @markpundsack what do you think?

dzaporozhets commented:
I think I will proceed with 3 commands suggestion from my comment above. Since buildpack has `test` and `test-compile` commands it might make sense to have same commands in herokuish + 1 command to execute both.

markpundsack commented:
Thanks, that sounds like a good plan.

michaelshobbs commented:
From the docs, it seems like the point of the `test` command is to compile the app via the correct buildpack and execute tests. Given that, I would actually fall on the other side and say we should only expose the single `test` command that internally calls the `compile` function and executes the app tests. I don't see the need to expose `test-compile` to the CLI.

michaelshobbs commented on include/buildpack.bash:
```diff
@@ -151,28 +174,53 @@ buildpack-execute() {
-	fi
-	cd - > /dev/null
-
+move-build-to-app() {
```
let's move this up near `_envfile-parse()` and change the function name to `_move-build-to-app`. the pattern here is unexported functions begin with `_` and are grouped with other unexported functions.

michaelshobbs commented on include/buildpack.bash:
```diff
@@ -117,6 +117,29 @@ buildpack-setup() {
+	move-build-to-app
+}
+
+select-buildpack() {
```
similar comment here as with `move-build-to-app`

dzaporozhets commented:
@michaelshobbs thanks for feedback. I will update MR asap and ping you when done

dzaporozhets commented:
@michaelshobbs comments addressed. Please take a look

dzaporozhets commented:
I also updated test apps for Ruby and Clojure and enabled `buildpack test` tests for them.

dzaporozhets commented:
@michaelshobbs Summary: 
1. Updated code based on your code review
1. Added new method in `buildpacks/test` to test `herokuish buildpack test`
1. Added tests to sample clojure and ruby apps
1. Enable testing of `herokuish buildpack test` for clojure and ruby apps
1. Bumped clojure buildpack version because testing is fixed there.

dzaporozhets commented:
@michaelshobbs I rebased and added a feature to README and changelog. I see you added @stickler-ci bot for automatic code review. Do I need to address any of those?

michaelshobbs commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ All notable changes to this project will be documented in this file.
 ### Fixed
 
 ### Added
+- @dzaporozhets Add support for `buildpack test` command
```
Go ahead and remove this for now. We update the changelog on release.

dzaporozhets replied:
@michaelshobbs got it. fixed

michaelshobbs commented:
@dzaporozhets rebasing this is sort of a pain. i'm rebasing it locally now and will push up a new branch + PR

swistaczek commented:
Hey, is it possible to use only 'test-compile' task? I would like to install all dependencies separately and share them in other tasks during CI pipeline.

josegonzalez commented:
@swistaczek please file a new issue rather than commenting on a year-old pull request. Thank you.
